src/storage/mvcc/reader/reader.rs

```diff
@@ -278,12 +279,104 @@ impl<S: EngineSnapshot> MvccReader<S> {
         Ok(())
     }
 
-    pub fn load_in_memory_pessimisitic_lock_range<F>(
+    /// Scan all types of locks(pessimitic, prewrite) satisfying `filter`
+    /// condition from both in-memory pessimitic lock table and the storage
+    /// within [start_key, end_key) .
+    pub fn scan_locks<F>(
+        &mut self,
+        start_key: Option<&Key>,
+        end_key: Option<&Key>,
+        filter: F,
+        limit: usize,
+        source: ScanLockReadTimeSource,
+    ) -> Result<(Vec<(Key, Lock)>, bool)>
+    where
+        F: Fn(&Key, TxnLockRef<'_>) -> bool,
+    {
+        let (memory_locks, memory_has_remain) = self.load_in_memory_pessimistic_lock_range(
+            start_key,
+            end_key,
+            |k, l| filter(k, l.into()),
+            limit,
+            source,
+        )?;
+        if memory_locks.is_empty() {
+            return self.scan_locks_from_storage(start_key, end_key, &filter, limit);
+        }
+
+        let mut lock_cursor_seeked = false;
+        let mut next_pair_from_storage = || -> Result<Option<(Key, Lock)>> {
+            self.create_lock_cursor()?;
+            let cursor = self.lock_cursor.as_mut().unwrap();
+            if !lock_cursor_seeked {
+                let ok = match start_key {
+                    Some(x) => cursor.seek(x, &mut self.statistics.lock)?,
+                    None => cursor.seek_to_first(&mut self.statistics.lock),
+                };
+                if !ok {
+                    return Ok(None);
+                }
+                lock_cursor_seeked = true;
```

Comment:
If `!ok`, in each iteration `seek` is called once. It could be avoided if we know we've tried and found nothing.

Reply:
Good catch, a status variable is added to check if it's exausted, PTAL.